README.md
# animated_textformfields

A Flutter package that has a collection of animated (and some not so animated) TextFormFields to make UI development faster and more beautiful.

## Example

![](gif/animated_demo.gif)

## Descriptions

#### Animated TextFormField

The Animated TextFormField toggles between animatiion states based on wheter is has focus or not. When the textformfield has focus, the textformfield animates the color and the corners of the textformfield.

#### SlideIn TextFormField

The SlideIn TextFormField animates between the hint text showing up inside the text field when the text is empty, and above the textformfield when it is filled with text.

#### Custom TextFormField Lite

The Custom TextFormField is a textformfield that does not animate. It is be used to complete another animated TextFormField, but can also be used as a standalone version. 


## Sample

#### Animated TextFormField

```
AnimatedTextFormField(
    width: MediaQuery.of(context).size.width * 0.875,
    height: 48.0,
    inputType: TextInputType.text,
    hintText: "Name",
    controller: textEditingControllerAnimated,
    textStyle: TextStyle(
        color: Colors.black,
        fontSize: 16.0,
    ),
    focusNode: myFocusNode,
    cornerRadius: BorderRadius.circular(14.0),
)
```

#### SlideIn TextFormField

```
SlideInTextFormField(
    vsync: this,
    width: MediaQuery.of(context).size.width * 0.875,
    height: 48.0,
    inputType: TextInputType.text,
    fieldText: "Name",
    controller: textEditingControllerSlideIn,
    textStyle: TextStyle(
        color: Colors.black,
        fontSize: 16.0,
    ),
    cornerRadius: BorderRadius.circular(14.0),
)
```

#### Custom TextFormField Lite

```
CustomTextFormFieldLite(
    width: MediaQuery.of(context).size.width * 0.875,
    height: 48.0,
    inputType: TextInputType.text,
    hintText: "Name",
    controller: textEditingControllerLite,
    textStyle: TextStyle(
        color: Colors.black,
        fontSize: 16.0,
    ),
    cornerRadius: BorderRadius.circular(14.0),
)
```

## Author

<a class="github-button" href="https://github.com/thealphaversion" aria-label="Follow @thealphaversion on GitHub">Follow @thealphaversion</a>

<a class="github-button" href="https://www.linkedin.com/in/aditya-chakraborti/" aria-label="LinkedIn: thealphaversion">LinkedIn: @Aditya Chakraborti</a>


CHANGELOG.md
## [0.1.0] - 21/06/2020

Added initial animated TextFormField that toggles between animatiion states based on wheter is has focus or not.


## [0.2.0] - 21/06/2020

Added a custom TextFormField that does not animate. It will be used to complete another animated TextFormField, but can also be used as a standalone version.

## [1.0.0] - 21/06/2020

Added a slide_in TextFormField that animated between the hint text showing up inside the text field when the text is empty, and above the textformfield when it is filled with text.

## [1.0.1] - 22/06/2020

Updated homepage for the package

test/animated_textformfields_test.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter/material.dart';
import 'package:animated_textformfields/animated_textformfields.dart';

void main() {
  TextEditingController textEditingControllerAnimated =
      new TextEditingController();
  TextEditingController textEditingControllerLite = new TextEditingController();
  TextEditingController textEditingControllerSlideIn = new TextEditingController();
  FocusNode focusNode = new FocusNode();
  testWidgets('Animated TextFormField renders', (WidgetTester tester) async {
    await tester.pumpWidget(
      MaterialApp(
        title: 'Animated TextFormField Test',
        home: Scaffold(
          body: Container(
            child: AnimatedTextFormField(
              width: 200.0,
              height: 48.0,
              inputType: TextInputType.text,
              hintText: "Name",
              controller: textEditingControllerAnimated,
              textStyle: TextStyle(
                color: Colors.black,
                fontSize: 16.0,
              ),
              focusNode: focusNode,
              cornerRadius: BorderRadius.circular(14.0),
            ),
          ),
        ),
      ),
    );
  });

  testWidgets('Custom TextFormField Lite renders', (WidgetTester tester) async {
    await tester.pumpWidget(
      MaterialApp(
        title: 'Custom TextFormField Lite Test',
        home: Scaffold(
          body: Container(
            child: CustomTextFormFieldLite(
              width: 200.0,
              height: 48.0,
              inputType: TextInputType.text,
              hintText: "Name",
              controller: textEditingControllerLite,
              textStyle: TextStyle(
                color: Colors.black,
                fontSize: 16.0,
              ),
              cornerRadius: BorderRadius.circular(14.0),
            ),
          ),
        ),
      ),
    );
  });
}


lib/animated_textformfield/animated_textformfield.dart
import 'package:flutter/material.dart';
import 'package:animated_textformfields/definitions/callbacks.dart';

class AnimatedTextFormField extends StatefulWidget {
  // sets the duration for the animation
  // default is 500 milliseconds
  final Duration animationDuration;


  // sets the width of the textformfield
  final double width;


  // sets the height of the textformfield
  // default is set to 48.0
  final double height;


  // toogles between a shadow and a very light border
  // default is false
  final bool hasShadow;


  // background color of the textformfield
  // default color is #EEEEEE
  final Color backgroundColor;


  // the corner radius of the textformfield when it does not have focus
  // default is 16.0
  final BorderRadius cornerRadius;


  // the corner radius of the textformfield when it is focused
  // default is 8.0
  final BorderRadius focusedCornerRadius;


  // sets the text style of both the text and the hintText
  // default is an empty string
  final TextStyle textStyle;


  // sets the hint text
  // this is blank by default
  final String hintText;


  // validator is a function that returns a string
  // if the entered text is invalid
  // return null if it is valid
  final StringCallback validator;


  // toggles between obscuring text and keeping plain text
  // recommended for fields that take sensitive information like passwords
  // default is false
  final bool obscureText;


  // defines the TextInputType
  // default is TextInputType.text
  final TextInputType inputType;


  // defines the focusNode
  // is used to switch between the animated states
  // based on whether the textformfield has focus
  final FocusNode focusNode;


  // defines the TextEditingController
  final TextEditingController controller;


  // defines the TextInputAction
  // default is TextInputAction.done
  final TextInputAction textInputAction;


  // defines the color when the textformfield has focus
  // default is Colors.white
  final Color focusColor;


  // sets the cursor color
  // default is #0A2552
  final Color cursorColor;


  // these are generic textformfield parameters
  final int maxLength, minLines, maxLines;
  final bool autocorrect;
  final bool autofocus;
  final EdgeInsets margin;
  final ValueChanged<String> onChanged, onSubmitted;
  final bool enabled;
  final GestureTapCallback onTap;

  const AnimatedTextFormField(
      {Key key,
      this.animationDuration = const Duration(milliseconds: 500),
      @required this.width,
      @required this.focusNode,
      this.height = 48.0,
      this.inputType,
      this.controller,
      this.margin = const EdgeInsets.all(10),
      this.obscureText = false,
      this.backgroundColor = const Color(0xFFEEEEEE),
      this.cornerRadius = const BorderRadius.all(Radius.circular(16)),
      this.focusedCornerRadius = const BorderRadius.all(Radius.circular(8)),
      this.cursorColor = const Color(0xff0a2552),
      this.focusColor = Colors.white,
      this.hintText = "",
      this.hasShadow = false,
      this.textStyle,
      this.validator,
      this.autofocus = false,
      this.autocorrect = false,
      this.enabled = true,
      this.maxLength,
      this.maxLines,
      this.minLines,
      this.onChanged,
      this.onTap,
      this.textInputAction,
      this.onSubmitted})
      : assert(width != null),
        assert(focusNode != null),
        super(key: key);

  @override
  _AnimatedTextFormFieldState createState() => _AnimatedTextFormFieldState();
}

class _AnimatedTextFormFieldState extends State<AnimatedTextFormField> {
  @override
  Widget build(BuildContext context) {
    return Center(
      // animates the textformfield
      child: AnimatedContainer(
        width: widget.width,
        height: widget.height,
        margin: widget.margin,
        alignment: Alignment.center,
        decoration: BoxDecoration(
          boxShadow: widget.hasShadow
              ? [
                  BoxShadow(
                    color: Colors.grey,
                    blurRadius: 2,
                    spreadRadius: 1,
                  ),
                ]
              : [
                  BoxShadow(
                    color: Colors.black38,
                    // has the effect of softening the shadow
                    blurRadius: 1.0,
                    // has the effect of extending the shadow
                    spreadRadius: 0.5,
                    offset: Offset(
                      0.0,
                      0.0,
                    ),
                  )
                ],
                // gives the widget rounded corners
          borderRadius: widget.focusNode.hasFocus
              ? widget.focusedCornerRadius
              : widget.cornerRadius,
              // defines the color of the widget
          color: widget.focusNode.hasFocus
              ? widget.focusColor
              : widget.backgroundColor,
        ),
        duration: widget.animationDuration,
        child: Padding(
          padding: const EdgeInsets.symmetric(horizontal: 16.0),
          child: TextFormField(
            decoration: new InputDecoration.collapsed(
              hintText: widget.hintText,
              hintStyle: widget.textStyle,
            ),
            style: widget.textStyle,
            controller: widget.controller,
            validator: widget.validator,
            obscureText: widget.obscureText,
            keyboardType: widget.inputType == null
                ? TextInputType.text
                : widget.inputType,
            autofocus: widget.autofocus,
            autocorrect: widget.autocorrect,
            focusNode: widget.focusNode,
            enabled: widget.enabled,
            maxLength: widget.maxLength,
            maxLines: widget.maxLines,
            minLines: widget.minLines,
            onChanged: widget.onChanged,
            cursorColor: widget.cursorColor,
            onTap: () {
              if (widget.onTap != null) {
                widget.onTap();
              }
            },
            onFieldSubmitted: (t) {
              widget.onSubmitted(t);
            },
            textInputAction: widget.textInputAction == null
                ? TextInputAction.done
                : widget.textInputAction,
          ),
        ),
      ),
    );
  }
}


lib/slide_in_textformfield/slide_in_textformfield.dart
import 'package:animated_textformfields/definitions/callbacks.dart';
import 'package:flutter/material.dart';
import 'package:animated_textformfields/custom_textformfield_lite/custom_textformfield_lite.dart';
import 'package:animated_size_and_fade/animated_size_and_fade.dart';

class SlideInTextFormField extends StatefulWidget {
  // sets the duration for the animation
  // default is 500 milliseconds
  final Duration animationDuration;


  // sets the vsync
  // required
  final TickerProvider vsync;


  // sets the width of the textformfield
  final double width;


  // sets the height of the textformfield
  // default is set to 48.0
  final double height;


  // toogles between a shadow and a very light border
  // default is false
  final bool hasShadow;


  // background color of the textformfield
  // default color is #EEEEEE
  final Color backgroundColor;


  // the corner radius of the textformfield when it does not have focus
  // default is 16.0
  final BorderRadius cornerRadius;


  // sets the text style of both the text and the hintText
  // default is an empty string
  final TextStyle textStyle;


  // sets the hint text and the field text
  // this is required
  final String fieldText;


  // validator is a function that returns a string
  // if the entered text is invalid
  // return null if it is valid
  final StringCallback validator;


  // toggles between obscuring text and keeping plain text
  // recommended for fields that take sensitive information like passwords
  // default is false
  final bool obscureText;


  // defines the TextInputType
  // default is TextInputType.text
  final TextInputType inputType;


  // defines the focusNode
  final FocusNode focusNode;


  // defines the TextEditingController
  final TextEditingController controller;


  // defines the TextInputAction
  // default is TextInputAction.done
  final TextInputAction textInputAction;


  // defines the color when the textformfield has focus
  // default is Colors.white
  final Color focusColor;


  // sets the cursor color
  // default is #0A2552
  final Color cursorColor;


  // the maximum lines of text it will render
  final int maxLines;


  // sets the TextOverFlow 
  final TextOverflow textOverflow;


  // sets the padding
  // default is EdgeInsets.fromLTRB(32.0, 0.0, 84.0, 0.0),
  final EdgeInsets padding;

  // these are generic textformfield parameters
  final int maxLength, minLines;
  final bool autocorrect;
  final bool autofocus;
  final EdgeInsets margin;
  final ValueChanged<String> onChanged, onSubmitted;
  final bool enabled;
  final GestureTapCallback onTap;

  const SlideInTextFormField(
      {Key key,
      this.animationDuration = const Duration(milliseconds: 500),
      @required this.vsync,
      @required this.width,
      @required this.controller,
      @required this.fieldText,
      this.focusNode,
      this.height = 48.0,
      this.inputType,
      this.margin = const EdgeInsets.all(10),
      this.obscureText = false,
      this.backgroundColor = const Color(0xFFEEEEEE),
      this.cornerRadius = const BorderRadius.all(Radius.circular(16)),
      this.cursorColor = const Color(0xff0a2552),
      this.focusColor = Colors.white,
      this.hasShadow = false,
      this.textStyle,
      this.validator,
      this.autofocus = false,
      this.autocorrect = false,
      this.enabled = true,
      this.maxLength,
      this.maxLines,
      this.minLines,
      this.padding = const EdgeInsets.fromLTRB(32.0, 0.0, 84.0, 0.0),
      this.onChanged,
      this.textOverflow = TextOverflow.ellipsis,
      this.onTap,
      this.textInputAction,
      this.onSubmitted})
      : assert(width != null),
        assert(vsync != null),
        assert(controller != null),
        assert(fieldText != null),
        super(key: key);

  @override
  _SlideInTextFormFieldState createState() => _SlideInTextFormFieldState();
}

class _SlideInTextFormFieldState extends State<SlideInTextFormField> {
  @override
  Widget build(BuildContext context) {
    return AnimatedContainer(
      duration: widget.animationDuration,
      child: Column(
        children: [
          AnimatedSizeAndFade(
            vsync: widget.vsync,
            child: widget.controller.text != ""
                ? Padding(
                    padding: widget.padding,
                    child: Align(
                      alignment: Alignment.centerLeft,
                      child: Text(
                        widget.fieldText,
                        overflow: widget.textOverflow,
                        maxLines: widget.maxLines,
                      ),
                    ),
                  )
                : Container(),
          ),
          CustomTextFormFieldLite(
            width: widget.width,
            height: widget.height,
            inputType: widget.inputType == null
                ? TextInputType.text
                : widget.inputType,
            textStyle: widget.textStyle,
            hasShadow: widget.hasShadow,
            backgroundColor: widget.backgroundColor,
            cornerRadius: widget.cornerRadius,
            onSubmitted: widget.onSubmitted,
            hintText: widget.fieldText,
            controller: widget.controller,
            validator: widget.validator,
            obscureText: widget.obscureText,
            autofocus: widget.autofocus,
            autocorrect: widget.autocorrect,
            focusNode: widget.focusNode,
            enabled: widget.enabled,
            maxLength: widget.maxLength,
            maxLines: widget.maxLines,
            minLines: widget.minLines,
            onChanged: widget.onChanged,
            onTap: widget.onTap,
            textInputAction: widget.textInputAction == null ? TextInputAction.done : widget.textInputAction,
          ),
        ],
      ),
    );
  }
}


lib/custom_textformfield_lite/custom_textformfield_lite.dart
import 'package:flutter/material.dart';
import 'package:animated_textformfields/definitions/callbacks.dart';

// defining a stringcallback that will be used to display validation messages
// typedef StringCallback = String Function(String);

class CustomTextFormFieldLite extends StatefulWidget {
  // sets the duration for the animation
  // default is 500 milliseconds
  final Duration animationDuration;


  // sets the width of the textformfield
  final double width;


  // sets the height of the textformfield
  // default is set to 48.0
  final double height;


  // toogles between a shadow and a very light border
  // default is false
  final bool hasShadow;


  // background color of the textformfield
  // default color is #EEEEEE
  final Color backgroundColor;


  // the corner radius of the textformfield when it does not have focus
  // default is 16.0
  final BorderRadius cornerRadius;


  // the corner radius of the textformfield when it is focused
  // default is 8.0
  final BorderRadius focusedCornerRadius;


  // sets the text style of both the text and the hintText
  // default is an empty string
  final TextStyle textStyle;


  // sets the hint text
  // this is blank by default
  final String hintText;


  // validator is a function that returns a string
  // if the entered text is invalid
  // return null if it is valid
  final StringCallback validator;


  // toggles between obscuring text and keeping plain text
  // recommended for fields that take sensitive information like passwords
  // default is false
  final bool obscureText;


  // defines the TextInputType
  // default is TextInputType.text
  final TextInputType inputType;


  // defines the focusNode
  final FocusNode focusNode;


  // defines the TextEditingController
  final TextEditingController controller;


  // defines the TextInputAction
  // default is TextInputAction.done
  final TextInputAction textInputAction;


  // defines the color when the textformfield has focus
  // default is Colors.white
  final Color focusColor;


  // sets the cursor color
  // default is #0A2552
  final Color cursorColor;


  // these are generic textformfield parameters
  final int maxLength, minLines, maxLines;
  final bool autocorrect;
  final bool autofocus;
  final EdgeInsets margin;
  final ValueChanged<String> onChanged, onSubmitted;
  final bool enabled;
  final GestureTapCallback onTap;

  const CustomTextFormFieldLite(
      {Key key,
      this.animationDuration = const Duration(milliseconds: 500),
      @required this.width,
      this.focusNode,
      this.height = 48.0,
      this.inputType,
      this.controller,
      this.margin = const EdgeInsets.all(10),
      this.obscureText = false,
      this.backgroundColor = const Color(0xFFEEEEEE),
      this.cornerRadius = const BorderRadius.all(Radius.circular(16)),
      this.focusedCornerRadius = const BorderRadius.all(Radius.circular(8)),
      this.cursorColor = const Color(0xff0a2552),
      this.focusColor = Colors.white,
      this.hintText = "",
      this.hasShadow = false,
      this.textStyle,
      this.validator,
      this.autofocus = false,
      this.autocorrect = false,
      this.enabled = true,
      this.maxLength,
      this.maxLines,
      this.minLines,
      this.onChanged,
      this.onTap,
      this.textInputAction,
      this.onSubmitted})
      : assert(width != null),
        super(key: key);

  @override
  _CustomTextFormFieldLiteState createState() => _CustomTextFormFieldLiteState();
}

class _CustomTextFormFieldLiteState extends State<CustomTextFormFieldLite> {
  @override
  Widget build(BuildContext context) {
    return Center(
      child: Container(
        width: widget.width,
        height: widget.height,
        margin: widget.margin,
        alignment: Alignment.center,
        decoration: BoxDecoration(
          boxShadow: [
            BoxShadow(
              color: Colors.black38,
              blurRadius: 1.0,
              // has the effect of softening the shadow
              spreadRadius: 0.5,
              // has the effect of extending the shadow
              offset: Offset(
                0.0, // horizontal, move right 10
                0.0, // vertical, move down 10
              ),
            )
          ],
          borderRadius: widget.cornerRadius,
          color: widget.backgroundColor,
        ),
        child: Padding(
          padding: const EdgeInsets.symmetric(horizontal: 16.0),
          child: TextFormField(
            decoration: new InputDecoration.collapsed(
              hintText: widget.hintText,
              hintStyle: widget.textStyle,
            ),
            style: widget.textStyle,
            controller: widget.controller,
            validator: widget.validator,
            obscureText: widget.obscureText,
            keyboardType: widget.inputType,
            autofocus: widget.autofocus,
            autocorrect: widget.autocorrect,
            focusNode: widget.focusNode,
            enabled: widget.enabled,
            maxLength: widget.maxLength,
            maxLines: widget.maxLines,
            minLines: widget.minLines,
            onChanged: widget.onChanged,
            onTap: () {
              if (widget.onTap != null) {
                widget.onTap();
              }
            },
            onFieldSubmitted: (t) {
              widget.onSubmitted(t);
            },
            textInputAction: TextInputAction.done,
          ),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:animated_textformfields/animated_textformfields.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Animated TextFields Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.orange,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: DemoPage(title: 'Demo Home Page'),
    );
  }
}

class DemoPage extends StatefulWidget {
  DemoPage({Key key, this.title}) : super(key: key);
  final String title;

  @override
  _DemoPageState createState() => _DemoPageState();
}

class _DemoPageState extends State<DemoPage> with TickerProviderStateMixin {
  FocusNode myFocusNode = FocusNode();
  TextEditingController textEditingControllerAnimated =
      new TextEditingController();
  TextEditingController textEditingControllerLite = new TextEditingController();
  TextEditingController textEditingControllerSlideIn = new TextEditingController();

  @override
  void dispose() {
    textEditingControllerAnimated.dispose();
    textEditingControllerLite.dispose();
    textEditingControllerSlideIn.dispose();
    myFocusNode.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Column(
        crossAxisAlignment: CrossAxisAlignment.center,
        children: <Widget>[
          Padding(
            padding:
                const EdgeInsets.symmetric(horizontal: 8.0, vertical: 24.0),
            child: Text(
              'Animated TextFields Demo',
              style: TextStyle(
                fontSize: 24.0,
              ),
              maxLines: 3,
            ),
          ),
          Align(
            alignment: Alignment.centerLeft,
            child: Padding(
              padding: const EdgeInsets.fromLTRB(36.0, 12.0, 12.0, 4.0),
              child: Text(
                'AnimatedTextFormField',
                style: TextStyle(
                  fontSize: 18.0,
                ),
                maxLines: 1,
              ),
            ),
          ),
          // this example describes how to use AnimatedTextFormField
          AnimatedTextFormField(
            width: MediaQuery.of(context).size.width * 0.875,
            height: 48.0,
            inputType: TextInputType.text,
            hintText: "Name",
            controller: textEditingControllerAnimated,
            textStyle: TextStyle(
              color: Colors.black,
              fontSize: 16.0,
            ),
            focusNode: myFocusNode,
            cornerRadius: BorderRadius.circular(14.0),
          ),
          Align(
            alignment: Alignment.centerLeft,
            child: Padding(
              padding: const EdgeInsets.fromLTRB(36.0, 24.0, 12.0, 4.0),
              child: Text(
                'CustomTextFormFieldLite',
                style: TextStyle(
                  fontSize: 18.0,
                ),
                maxLines: 1,
              ),
            ),
          ),
          // this example describes how to use CustomTextFormFieldLite
          CustomTextFormFieldLite(
            width: MediaQuery.of(context).size.width * 0.875,
            height: 48.0,
            inputType: TextInputType.text,
            hintText: "Name",
            controller: textEditingControllerLite,
            textStyle: TextStyle(
              color: Colors.black,
              fontSize: 16.0,
            ),
            cornerRadius: BorderRadius.circular(14.0),
          ),
          Align(
            alignment: Alignment.centerLeft,
            child: Padding(
              padding: const EdgeInsets.fromLTRB(36.0, 24.0, 12.0, 4.0),
              child: Text(
                'SlideInTextFormField',
                style: TextStyle(
                  fontSize: 18.0,
                ),
                maxLines: 1,
              ),
            ),
          ),
          // this example describes how to use SlideInTextFormField
          SlideInTextFormField(
            vsync: this,
            width: MediaQuery.of(context).size.width * 0.875,
            height: 48.0,
            inputType: TextInputType.text,
            fieldText: "Name",
            controller: textEditingControllerSlideIn,
            textStyle: TextStyle(
              color: Colors.black,
              fontSize: 16.0,
            ),
            cornerRadius: BorderRadius.circular(14.0),
          ),
        ],
      ),
    );
  }
}
